restapi/src/main/java/io/stargate/web/docsapi/service/query/search/resolver/CnfResolver.java

```diff
@@ -0,0 +1,237 @@
+
+import com.bpodgursky.jbool_expressions.And;
+import com.bpodgursky.jbool_expressions.Expression;
+import com.bpodgursky.jbool_expressions.Literal;
+import com.bpodgursky.jbool_expressions.options.ExprOptions;
+import com.bpodgursky.jbool_expressions.rules.Rule;
+import com.bpodgursky.jbool_expressions.rules.RuleList;
+import com.bpodgursky.jbool_expressions.rules.RulesHelper;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import io.stargate.web.docsapi.service.ExecutionContext;
+import io.stargate.web.docsapi.service.query.FilterExpression;
+import io.stargate.web.docsapi.service.query.FilterPath;
+import io.stargate.web.docsapi.service.query.rules.TrueFilterExpressions;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.CandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.impl.InMemoryCandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.filter.impl.PersistenceCandidatesFilter;
+import io.stargate.web.docsapi.service.query.search.resolver.impl.AllFiltersResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.impl.InMemoryDocumentsResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.impl.PersistenceDocumentsResolver;
+import io.stargate.web.docsapi.service.query.search.weigth.ExpressionWeightResolver;
+import io.stargate.web.docsapi.service.query.search.weigth.impl.UserOrderWeightResolver;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/** Static resolver that accepts the {@link And} expression in the CNF form. */
+public final class CnfResolver {
+
+  private CnfResolver() {}
+
+  /**
+   * Returns a document resolver for a single {@link And} expression in CNF form without a parent.
+   *
+   * @param expression {@link FilterExpression}
+   * @param context {@link ExecutionContext}
+   * @return DocumentsResolver
+   */
+  public static DocumentsResolver resolve(
+      Expression<FilterExpression> expression, ExecutionContext context) {
+    return resolve(expression, context, null);
+  }
+
+  /**
+   * Returns a document resolver for a single {@link And} expression in CNF form.
+   *
+   * @param expression {@link FilterExpression}
+   * @param context {@link ExecutionContext}
+   * @param parent parent resolver or <code>null</code>
+   * @return DocumentsResolver
+   */
+  public static DocumentsResolver resolve(
+      Expression<FilterExpression> expression, ExecutionContext context, DocumentsResolver parent) {
+    // from the children inside and
+    And<FilterExpression> andExpression = (And<FilterExpression>) expression;
+    List<Expression<FilterExpression>> children = andExpression.getChildren();
+
+    UserOrderWeightResolver weightResolver = new UserOrderWeightResolver(children);
```

Comment:
We should probably construct this before converting to CNF.

Reply:
Yep this should be done. But I like this idea of assigning selectivity value, then we can make this completely static.

Reply:
Moved to order and static `UserOrderWeightResolver`